1-Authentication/7-sign-in-express-mfa/App/auth/AuthProvider.js

```diff
@@ -0,0 +1,253 @@
+const axios = require('axios');
+const { msalConfig, TENANT_SUBDOMAIN, REDIRECT_URI, POST_LOGOUT_REDIRECT_URI } = require('../authConfig');
+
+class AuthProvider {
+    config;
+    cryptoProvider;
+
+    constructor(config) {
+        this.config = config;
+        this.cryptoProvider = new msal.CryptoProvider();
+    }
+
+    getMsalInstance(msalConfig) {
+        return new msal.ConfidentialClientApplication(msalConfig);
+    }
+
+    async login(req, res, next, options = {}) {
+        // create a GUID for crsf
+        req.session.csrfToken = this.cryptoProvider.createNewGuid();
+
+        /**
+         * The MSAL Node library allows you to pass your custom state as state parameter in the Request object.
+         * The state parameter can also be used to encode information of the app's state before redirect.
+         * You can pass the user's state in the app, such as the page or view they were on, as input to this parameter.
+         */
+        const state = this.cryptoProvider.base64Encode(
+            JSON.stringify({
+                csrfToken: req.session.csrfToken,
+                redirectTo: '/',
+            })
+        );
+
+        const authCodeUrlRequestParams = {
+            state: state,
+
+            /**
+             * By default, MSAL Node will add OIDC scopes to the auth code url request. For more information, visit:
+             * https://docs.microsoft.com/azure/active-directory/develop/v2-permissions-and-consent#openid-connect-scopes
+             */
+            scopes: options.scopes ?? [],
+        };
+
+        const authCodeRequestParams = {
+            state: state,
+
+            /**
+             * By default, MSAL Node will add OIDC scopes to the auth code request. For more information, visit:
+             * https://docs.microsoft.com/azure/active-directory/develop/v2-permissions-and-consent#openid-connect-scopes
+             */
+            scopes: options.scopes ?? [],
+        };
+
+        /**
+         * If the current msal configuration does not have cloudDiscoveryMetadata or authorityMetadata, we will
+         * make a request to the relevant endpoints to retrieve the metadata. This allows MSAL to avoid making
+         * metadata discovery calls, thereby improving performance of token acquisition process.
+         */
+        if (!this.config.msalConfig.auth.authorityMetadata) {
+            const authorityMetadata = await this.getAuthorityMetadata();
+            this.config.msalConfig.auth.authorityMetadata = JSON.stringify(authorityMetadata);
+        }
+
+        const msalInstance = this.getMsalInstance(this.config.msalConfig);
+
+        // trigger the first leg of auth code flow
+        return this.redirectToAuthCodeUrl(
+            req,
+            res,
+            next,
+            authCodeUrlRequestParams,
+            authCodeRequestParams,
+            msalInstance
+        );
+    }
+
+    doesRequireMFA(token) {
+        // Decode the access token payload
```

Comment:
I don't think this is a pattern we want to demonstrate - parsing access tokens will not work when token encryption is enabled, rendering access tokens opaque. Instead, I think we implement the sample such that  the client doesn't need to know if the call requires MFA or not and just "does the right thing" when MFA is required.

Reply:
Removed the code that did this.
Updated the flow you suggested.